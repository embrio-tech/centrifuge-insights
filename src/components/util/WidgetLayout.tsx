import React, { useRef } from 'react'
import { Responsive as ResponsiveGridLayout } from 'react-grid-layout'
import { gridBreakpoints, gridCols, gridMargin } from '../../config'
// import './WidgetLayout.less'
import 'react-grid-layout/css/styles.css'
import 'react-resizable/css/styles.css'
import { useSize, useWidgetLayout } from '../../hooks'
import { WidgetAppearance } from '../../types'

interface WidgetLayoutProps {
  className?: string
  widgets: WidgetAppearance[]
  edit?: boolean
}

const WidgetLayout: React.FC<WidgetLayoutProps> = (props) => {
  const { className, children, widgets, edit = false } = props
  const ref = useRef(null)
  const { width } = useSize(ref)

  const layouts = useWidgetLayout(widgets)

  return (
    <div ref={ref} className={className}>
      <ResponsiveGridLayout
        width={width}
        className='layout'
        layouts={layouts}
        breakpoints={gridBreakpoints}
        cols={gridCols}
        margin={gridMargin}
        isDraggable={edit}
        isResizable={edit}
        // onBreakpointChange={(breakpoint: string, cols: number) => {
        //   // TODO: handle breakpoint change
        //   // console.log('breakpoint: ', breakpoint, cols)
        // }}
        // onLayoutChange={(layout, allLayouts) => {
        //   // TODO: handle layout change
        //   // console.log('layout: ', layout)
        //   // console.log('allLayouts: ', allLayouts)
        // }}
      >
        {children}
      </ResponsiveGridLayout>
    </div>
  )
}

export { WidgetLayout }
