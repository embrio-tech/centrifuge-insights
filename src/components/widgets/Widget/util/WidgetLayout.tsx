import React, { ReactNode } from 'react'
import './WidgetLayout.less'
import { WidgetSpin } from './WidgetSpin'

interface WidgetLayoutProps {
  className?: string
  loading?: boolean
  header?: ReactNode
  footer?: ReactNode
}

/**
 * generic widget layout component with header, body, footer, and laoding spinner
 *
 * @prop {ReactNode} children - body component(s)
 * @prop {ReactNode?} header - header component (optional)
 * @prop {ReactNode?} footer - footer component (optional)
 * @prop {boolean?} loading - flag to show loading spinner (optional, default: false)
 * @prop {string?} className - class names applied to wrapping <div> (optional)
 */
export const WidgetLayout: React.FC<WidgetLayoutProps> = (props) => {
  const { className, children, loading = false, header, footer } = props

  return (
    <div className={className}>
      <WidgetSpin spinning={loading}>
        <div className='h-full flex flex-col overflow-hidden'>
          {header && <div className='grow-0'>{header}</div>}
          <div className='grow h-px'>
              <div className='h-full'>{children}</div>
          </div>
          {footer && <div className='grow-0'>{footer}</div>}
        </div>
      </WidgetSpin>
    </div>
  )
}
