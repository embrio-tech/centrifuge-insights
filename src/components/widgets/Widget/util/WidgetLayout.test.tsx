import React from 'react'
import renderer from 'react-test-renderer'
import { WidgetLayout } from './WidgetLayout'

describe('WidgetLayout should', () => {
  test('match snapshot!', () => {
    const tree = renderer
      .create(
        <WidgetLayout className='foo-class' header='This is the header!' footer='This is the footer!'>
          This is the body!
        </WidgetLayout>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

export default null
