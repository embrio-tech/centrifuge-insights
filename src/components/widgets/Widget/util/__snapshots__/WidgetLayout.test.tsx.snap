// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetLayout should match snapshot! 1`] = `
<div
  className="foo-class"
>
  <div
    className="ant-spin-nested-loading widget-spin"
  >
    <div
      className="ant-spin-container"
    >
      <div
        className="h-full flex flex-col overflow-hidden"
      >
        <div
          className="grow-0"
        >
          This is the header!
        </div>
        <div
          className="grow h-px"
        >
          <div
            className="h-full"
          >
            This is the body!
          </div>
        </div>
        <div
          className="grow-0"
        >
          This is the footer!
        </div>
      </div>
    </div>
  </div>
</div>
`;
